storage/src/main/java/org/apache/uniffle/storage/handler/impl/DataSkippableReadHandler.java

```diff
@@ -59,8 +60,17 @@ public DataSkippableReadHandler(
 
   public ShuffleDataResult readShuffleData() {
     if (shuffleDataSegments.isEmpty()) {
-      ShuffleIndexResult shuffleIndexResult = readShuffleIndex();
+      ShuffleIndexResult shuffleIndexResult;
+      try {
+        shuffleIndexResult = readShuffleIndex();
+      } catch (Exception e) {
+        if (++failTimes > maxHanderFailTimes) {
+          isFinished = true;
+        }
+        throw e;
```

Comment:
This exception would be handled by`HdfsClientReadHandler`, which increases the `handlerIndex`  and moves to the next handler, so what is the point of the `finished` flag here?

Reply:
`isFinished` is for mark don't try read from this handler.  In original logic, if `HdfsClientReadHandler` catch this exception, it will not retry read from this handler, but it will retry in this pr.

Reply:
`HdfsClientReadHandler` catches this exception and moves forward the `readHandlerIndex ` according to the diff in `HdfsClientReadHandler`.

Reply:
Yes, if the first `hdfsShuffleFileReader` fail, it will read the next one. But how about all `hdfsShuffleFileReader` fail for network temporarily unavailable. Do i get your point?

Reply:
I mean according to your design the handle will be used again before reaching the `maxHanderFailTimes`, but the readHandlerIndex increased in the exception handling block, how can it be used again? Are there more rounds from the upper layer handler to traverse the handlers from index 0?

Reply:
Yes,you can search `nextRound` in this pr

Reply:
So it may loop the handlers a second time, skip the finished ones and some handlers may be still unfinished for the maxHandlerFailTimes might be greater than 2? 
I have two suggestions, could you help to review?
- Do the retry in the upper layer and only keep state data such as the finish flag and `segmentIndex` in the low-level handlers, and derive the finish flag from the `segmentIndex`, which can be also used as a checkpoint for retry.
- **Do Not** expose the retry logic to the `ShuffleClient`, make it cohesive in the `ClientReadHandler` and can be configured.

Reply:
> So it may loop the handlers a second time, skip the finished ones and some handlers may be still unfinished for the maxHandlerFailTimes might be greater than 2?

Yes, but what's the problem and what's the impact？I don't understand.

> * Do the retry in the upper layer

Upper layer means `ShuffleClient`? I think we must retry in `ShuffleClient`, because we don't need read from all handler in every round, and we need to check blocks consistent every round

Reply:
>Yes, but what's the problem and what's the impact？I don't understand.

Not a problem, we do not need to record the retries in the low-level handlers

>Upper layer means ShuffleClient? I think we must retry in ShuffleClient

Yes, agree. Check consistency and call the reshuffleData in `ShuffleClient`, and do the actual handlers retries (low-level handler loops) in the `clientReadHandler`.

Reply:
If the first server stop, should we retry in the handler in the first round? In most cases, handler fail is for shuffle server is unavailable, we should try anther handler right now in this case.

Reply:
I think retry in the rpc client is better than retry in handler